### ISSUE TITLE
MP4Box doesn't work well with non ASCII path: the remove file feature doesn't work

### ISSUE THREAD

**Issue by CyberSinh** [gpac/gpac#883]
If I type the following command:

`"C:\Program Files\GPAC\mp4box.exe" -par 1=1:1 "E:\NAS\video\Vidéos de famille\2010\Alex 2010\HDV_0091.MP4"`

in a command prompt (cmd.exe) under Windows 10, the temp file "out_HDV_0091.MP4" is correclty created, but the command cannot finish because there is an error when the soft try to remove the old file:

```
Saving E:\NAS\video\Vid├®os de famille\2010\Alex 2010\HDV_0091.MP4: 0.500 secs Interleaving
Error removing file E:\NAS\video\Vid├®os de famille\2010\Alex 2010\HDV_0091.MP4
Error: I/O Error
```

I get the same error with Powershell:

```
Saving E:\NAS\video\VidÃ©os de famille\2010\Alex 2010\HDV_0091.MP4: 0.500 secs Interleaving
Error removing file E:\NAS\video\VidÃ©os de famille\2010\Alex 2010\HDV_0091.MP4
Error: I/O Error
```

The issue seems to be related with the space and the 'é' char contained in path. Perhaps the remove function called to drop the file doesn't support Unicode path.

As workaround, I use the "-out" parameter to not use the in-place rewrite feature. I manually delete old video files after processing.

**Comment by rbouqueau** [comment]
Windows 10 here, and it seems to work in 'cmd':
```
C:\Users\rbouqueau>MP4Box -par 1=1:1 -add "D:\Works\gpac\issues\883_unicode\Vidéos de famille\counter_video_180.mp4" -new "D:\Works\gpac\issues\883_unicode\Vidéos de famille\tmp.mp4"
IsoMedia import counter_video_180.mp4 - track ID 1 - Video (size 320 x 180)
Saving D:\Works\gpac\issues\883_unicode\Vid├®os de famille\tmp.mp4: 0.500 secs Interleaving

C:\Users\rbouqueau>
```

Are you using the latest version of GPAC? What other information could you provide us so that we can reproduce?

**Comment by CyberSinh** [comment]
I use the latest stable version (x64) on Windows 10.

The 'new' parameter was not used.
The input file was not on the same partition than the temp one.

**Comment by rbouqueau** [comment]
That's not the first time users report this issue @jeanlf @aureliendavid  . We usually tell to specify the ```-tmp``` dir manually but may need a more automatic solution.

**Comment by CyberSinh** [comment]
I tried to use the 'tmp' parameter to set a temp dir on the same partition than the input files, but without any effect on the issue.

**Comment by rbouqueau** [comment]
The issue is that the MoveFile() function from the Windows API fails when on different drives. And the 'tmp' directory is retrieved with the tmpfile() function (from the compiler runtime, and we use Microsoft's one in our automatic builds) ; very convenient when people wish to write to slow drives ; however combination of both functions above trigger this issue.

**Comment by CyberSinh** [comment]
I tried to use the tmp parameter to set input and temp files on the same drive, but the issue "error removing file" remained.

**Comment by aureliendavid** [comment]
I can't reproduce the issue about temp and destination files being on different drives. It seems to work :
`
C:\Users\adavid>"C:\Program Files\GPAC\mp4box.exe" -par 1=1:1 "D:\PANA.mp4"`

no problem there. 

However, it only works if the permissions on the destination folder/file are correct. And windows permissions are weird.

For example, if I try to write to program files it will fail because we need an elevated prompt (admin) to do that. So 

`"C:\Program Files\GPAC\mp4box.exe" -par 1=1:1 "C:\Program Files\GPAC\PANA.mp4"`

fails. It also fails if the destination file or its parent folder is set **read-only** : 

`C:\Users\adavid>"C:\Program Files\GPAC\mp4box.exe" -par 1=1:1 "D:\ro\PANA.mp4"`

This fails ***even though I can delete the file in explorer as normal user***. 

I see that in the original report the destination started with "E:\NAS\" so I assume there is some samba or nfs mount behind, which opens a whole other can of worms in terms of windows permissions. 

So I'd say try on a destination file that you are sure you can delete. You can test this by **first making a copy of the file** and then trying the `del <path>` command from the same prompt to see if it's a windows issue or an mp4box issue.

**Comment by CyberSinh** [comment]
Despite its name, E:\NAS is a path located on a fixed HDD.
To avoid this issue, I used the new parameter to write new file before removing old files by hand.

**Comment by aureliendavid** [comment]
Can you still try to backup the file that MP4Box can't write to in-place, and try to delete it (the original, not the backup) in a command prompt with the `del` command? That way we can know if there's a bug in MP4Box or if it's just windows being windows. 

Thanks.

**Comment by CyberSinh** [comment]
It doesn't seem to be a file permission issue. Here is a screencast showing the issue:

![mp4box_move_issue](https://user-images.githubusercontent.com/10407531/29046792-4ee54b84-7bc9-11e7-8942-7a45b9f95cce.gif)

Do not hesitate if I can do anything to help you to diagnose the bug.

**Comment by aureliendavid** [comment]
great gif, that helps :)

**TL;DR:** it turns out you were right from the start, it is a problem with non-ascii paths (sorry about that)

can you retry with this version: https://download.tsi.telecom-paristech.fr/gpac/latest_builds/win64/gpac-0.7.2-DEV-rev225-gbf91c8a-master-x64.exe?


**long version for the record:**

Windows passes its arguments to `main()` as wide strings (`wchar_t *`). Linux passes multibyte strings (utf8 `char*`). To use the same code everywhere, on windows we start by converting `wchar_t*` arguments to utf8 `char*` arguments. 

**Big however**, it looks like Windows doesn't actually supports utf8 multibyte `char*` strings, but uses its own codepages (e.g. `windows-1252`). This is why `printf`-ing a path with an accent prints bad characters, even when mp4box works. (See for example https://msdn.microsoft.com/en-us/library/883tf19a.aspx *"except UTF-7 and UTF-8, which are not supported"*).

The reason mp4box works with input files and -new is because `gf_fopen()` has a fallback where we convert the path back to `wchar_t*` is we can't open it as `char*`. But the other file handling functions don't have this fallback, including `gf_delete_file()`. 

So I replaced the conversion of arguments to the current active codepage instead of utf8 and it seems to work. Hopefully this doesn't break something else...

**Comment by CyberSinh** [comment]
It works now!
Merci beaucoup.

**Comment by sverrirs** [comment]
When will this fix be released on https://gpac.wp.imt.fr/downloads/? Having these same problems and would like to redist mp4box with my scripts.

**Comment by aureliendavid** [comment]
It's been included in published nightly installers (https://gpac.wp.imt.fr/downloads/gpac-nightly-builds/) since the issue was closed. Are you having charset troubles with recent installers? 

The solution to this issue was superseded by https://github.com/gpac/gpac/pull/919 which is also included in recent releases. 

Nightly builds should be as stable as "big" releases so you can use these. I'm not sure when the next version change release will be.

**Comment by sverrirs** [comment]
Yeah I pulled the latest nightly before and it works great. Just wanted to ask due to the redistribution restrictions on the dev builds :) 

>  These installers are distributed for educational purposes only, and may not be redistributed.

**Comment by rbouqueau** [comment]
@sverrirs I don't remember why I wrote this sentence in 2011, but you'll be happy to know that I've just removed it (in 2018)!

**Comment by sverrirs** [comment]
Merci Romain :)
